### ISSUE TITLE
ask feature : save page in incognito mode

### ISSUE THREAD

**Issue by Ananas-Kuchen** [bromite/bromite#1182]
### Is your feature request related to privacy?

 yes

### Is there a patch available for this feature somewhere?

 no

### Describe the solution you would like

 please add option to save web pages in incognito mode. because I really need it.

### Describe alternatives you have considered

 no idea

**Comment by csagan5** [comment]
@uazo you have also solved this with your PR #1165, so this should be fixed with next release.

**Comment by uazo** [comment]
no, what I re-enabled is long-press -> download link

unfortunately I didn't even know about that feature.

![image](https://user-images.githubusercontent.com/29201891/121484516-96c77780-c9cf-11eb-8914-26f04dbaaa3e.png)

it is part of those things to be rehabilitated in incognito (always and "normal").

**Comment by csagan5** [comment]
Ok; so I will keep this open.

**Comment by uazo** [comment]
>  (always and "normal").

what do you think, only for always-incognito or in general for the all incognito modes?

**Comment by csagan5** [comment]
> what do you think, only for always-incognito or in general for the all incognito modes?

All modes; I do not understand why it is not available currently, could it introduce any privacy problem?

**Comment by Ananas-Kuchen** [comment]
Can someone give me a hint how to save web page (mhtml) in incognito mode as mentioned in this latest version changelog?

https://github.com/bromite/bromite/releases/tag/91.0.4472.102



> no, what I re-enabled is long-press -> download link
> 
> unfortunately I didn't even know about that feature.
> 
> ![image](https://user-images.githubusercontent.com/29201891/121484516-96c77780-c9cf-11eb-8914-26f04dbaaa3e.png)
> 
> it is part of those things to be rehabilitated in incognito (always and "normal").

I experienced thing like in this screenshot. I only found the download button can be used in non-incognito mode

If there is no way. alright. incognito mode doesn't help me with this. 😔

**Comment by uazo** [comment]
> All modes; I do not understand why it is not available currently, could it introduce any privacy problem?

an accidental press of the button produces the page on disk without further request to the user, who instead does not want to leave a trace.
a confirmation should be added.

> Can someone give me a hint how to save web page (mhtml) in incognito mode as mentioned in this latest version changelog?

@Ananas-Kuchen it is not there yet, it is in development

**Comment by csagan5** [comment]
> > All modes; I do not understand why it is not available currently, could it introduce any privacy problem?
> 
> an accidental press of the button produces the page on disk without further request to the user, who instead does not want to leave a trace.
> a confirmation should be added.

I disagree; the page is already downloaded in the cache of your device. This would be a further copy; there is an UI indication of the download starting and you can go there and remove it. Unless you consider a trace the downloading itself, but how is that different from clicking a hyperlink?

**Comment by uazo** [comment]
> I disagree; the page is already downloaded in the cache of your device

the cache is ephemeral and is not persisted to disk

> This would be a further copy; there is an UI indication of the download starting and you can go there and remove it. 

sure. but it is possible to undo the delete, I don't know where that information is stored

> but how is that different from clicking a hyperlink?

with clicking a hyperlink nothing is saved to disk...
But since it seems like an obvious answer to me, it's easy that I don't understand well your question..

**Comment by csagan5** [comment]
> > I disagree; the page is already downloaded in the cache of your device
> 
> the cache is ephemeral and is not persisted to disk

What does get saved to disk in incognito mode? Only sessions?

> > but how is that different from clicking a hyperlink?
> 
> with clicking a hyperlink nothing is saved to disk...
> But since it seems like an obvious answer to me, it's easy that I don't understand well your question..

I thought that pages get saved to cache, but the cache itself is a file stored on disk...not fully on RAM.

If we have certainty that navigation pages are only stored to RAM, how can this feature be implemented? A prompt for incognito pages only?

**Comment by uazo** [comment]
> What does get saved to disk in incognito mode? Only sessions?

yes the sessions are encrypted on disk, nothing else.

>  thought that pages get saved to cache, but the cache itself is a file stored on disk...not fully on RAM.

when I had fixed the always incognito patch the first time, I had seen that nothing is saved on disk, but I don't know how android/linux save the memory on swap file, in windows it can be said not to do it, in android never investigated. and also I don't know how java behaves with memory

> A prompt for incognito pages only?

Sì

**Comment by csagan5** [comment]
There is no swap on Android; I will leave this feature assigned to you, you can implement it with the prompt when you want/have time.

**Comment by uazo** [comment]
> There is no swap on Android

Sure? i don't know anything about android but zram what is it then?

https://source.android.com/devices/tech/perf/low-ram
https://android.googlesource.com/kernel/msm/+/android-7.1.0_r0.2/Documentation/blockdev/zram.txt

**Comment by csagan5** [comment]
You can search about it; it is not swap, that documentation page explains it as well.

**Comment by uazo** [comment]
> I do not understand why it is not available currently, could it introduce any privacy problem?

now, only after developing it can I give you an answer.

the feauture is disabled in incognito because it keeps a db of the downloaded pages on disk (https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/offline_pages/android/offline_page_model_factory.cc;l=64;drc=2e2aee0550b8ff482c0329bad62f71c49407d40a;bpv=1;bpt=1), in addition to the mhtml file, for some metadata.
furthermore, only in normal navigation it is possible to use other features related to offline pages (such as scheduled auto fetch).

in incognito browsing it is only expected to be able to open an mhtml file, the rest I was unable to reactivate it and perhaps it is better this way.

moreover, the patch is rather invasive because it reactivates in incognito browsing some services that have been deliberately (and not recently https://source.chromium.org/chromium/chromium/src/+/4727ab67dde1088bdb5da8c63c51ee91e929e6e9) disabled, unlike other services however active in incognito mode. I mean invasive because only if in the latter case, some ifs with `isIncognito()` are expected.

so, the first version I will send you only reactivates them in always_incognito, what do you think?
